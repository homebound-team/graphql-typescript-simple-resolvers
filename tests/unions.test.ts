import { createTestSchema, runPlugin } from "./utils/test-helpers";

describe("Union Types", () => {
  const schemaWithUnions = `
    type Author {
      id: ID!
      name: String!
    }

    type Book {
      id: ID!
      title: String!
    }

    type Article {
      id: ID!
      headline: String!
    }

    # Basic union of object types
    union SearchResult = Author | Book

    # Union including primitives
    union StringOrNumber = String | Int

    # Union with primitives and object types
    union MixedUnion = String | Boolean | Author

    # Nested union (union of unions)
    union UnionOfUnions = SearchResult | StringOrNumber

    type Query {
      search(query: String!): [SearchResult!]!
      mixedSearch: MixedUnion
      testNestedUnion: UnionOfUnions
      getPrimitive: StringOrNumber
    }

    # Union in nullable field
    type Container {
      optionalResult: SearchResult
      requiredResult: SearchResult!
      optionalResults: [SearchResult!]
      requiredResults: [SearchResult!]!
    }
  `;

  it("generates basic union types", async () => {
    const schema = createTestSchema(schemaWithUnions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain("export type SearchResult = Author | Book;");
  });

  it("generates union with primitives", async () => {
    const schema = createTestSchema(schemaWithUnions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain("export type StringOrNumber = string | number;");
    expect(code).toContain("export type MixedUnion = string | boolean | Author;");
  });

  it("generates nested unions", async () => {
    const schema = createTestSchema(schemaWithUnions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain("export type UnionOfUnions = SearchResult | StringOrNumber;");
  });

  it("generates union resolvers for object types", async () => {
    const schema = createTestSchema(schemaWithUnions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`SearchResult: { __resolveType(o: Author | Book): string };`);
  });

  it("generates union resolvers for mixed unions", async () => {
    const schema = createTestSchema(schemaWithUnions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`MixedUnion: { __resolveType(o: String | Boolean | Author): string };`);
  });

  it("handles unions in field types with correct nullability", async () => {
    const schema = createTestSchema(schemaWithUnions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface Container {
  optionalResult: SearchResult | null | undefined;
  requiredResult: SearchResult;
  optionalResults: SearchResult[] | null | undefined;
  requiredResults: SearchResult[];
}`);
  });

  it("handles unions in query resolvers", async () => {
    const schema = createTestSchema(schemaWithUnions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`search: Resolver<{}, QuerySearchArgs, readonly SearchResult[]>;`);
    expect(code).toContain(`mixedSearch: Resolver<{}, {}, MixedUnion | null | undefined>;`);
  });

  it("snapshots union type generation", async () => {
    const schema = createTestSchema(schemaWithUnions);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });
    expect(code).toMatchSnapshot();
  });

  it("handles union property nullability correctly", async () => {
    const schemaWithNullableUnion = `
      type Author {
        name: String!
      }

      type Book {
        title: String!
      }

      union Content = Author | Book

      type Post {
        content: Content
        requiredContent: Content!
      }

      type Query {
        posts: [Post!]!
      }
    `;

    const schema = createTestSchema(schemaWithNullableUnion);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface Post {
  content: Content | null | undefined;
  requiredContent: Content;
}`);
  });
});
