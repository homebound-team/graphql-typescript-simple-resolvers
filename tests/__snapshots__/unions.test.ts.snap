// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Union Types snapshots union type generation 1`] = `
"import { GraphQLResolveInfo } from "graphql";
import { Context } from "./context";

export interface Resolvers {
  Query: QueryResolvers;
  Author?: AuthorResolvers;
  Book?: BookResolvers;
  Article?: ArticleResolvers;
  Container?: ContainerResolvers;
}

export type UnionResolvers = {
  SearchResult: { __resolveType(o: Author | Book): string };
  StringOrNumber: { __resolveType(o: String | Int): string };
  MixedUnion: { __resolveType(o: String | Boolean | Author): string };
  UnionOfUnions: { __resolveType(o: SearchResult | StringOrNumber): string };
};

export interface QueryResolvers {
  search: Resolver<{}, QuerySearchArgs, readonly SearchResult[]>;
  mixedSearch: Resolver<{}, {}, MixedUnion | null | undefined>;
  testNestedUnion: Resolver<{}, {}, UnionOfUnions | null | undefined>;
  getPrimitive: Resolver<{}, {}, StringOrNumber | null | undefined>;
}

export interface AuthorResolvers {
  id: Resolver<Author, {}, string>;
  name: Resolver<Author, {}, string>;
}

export interface BookResolvers {
  id: Resolver<Book, {}, string>;
  title: Resolver<Book, {}, string>;
}

export interface ArticleResolvers {
  id: Resolver<Article, {}, string>;
  headline: Resolver<Article, {}, string>;
}

export interface ContainerResolvers {
  optionalResult: Resolver<Container, {}, SearchResult | null | undefined>;
  requiredResult: Resolver<Container, {}, SearchResult>;
  optionalResults: Resolver<Container, {}, readonly SearchResult[] | null | undefined>;
  requiredResults: Resolver<Container, {}, readonly SearchResult[]>;
}

type MaybePromise<T> = T | Promise<T>;
export type Resolver<R, A, T> = (root: R, args: A, ctx: Context, info: GraphQLResolveInfo) => MaybePromise<T>;

export type SubscriptionResolverFilter<R, A, T> = (
  root: R | undefined,
  args: A,
  ctx: Context,
  info: GraphQLResolveInfo,
) => boolean | Promise<boolean>;
export type SubscriptionResolver<R, A, T> = {
  subscribe: (root: R | undefined, args: A, ctx: Context, info: GraphQLResolveInfo) => AsyncIterator<T>;
};

export interface QuerySearchArgs {
  query: string;
}
export interface Author {
  id: string;
  name: string;
}

export interface Book {
  id: string;
  title: string;
}

export interface Article {
  id: string;
  headline: string;
}

export interface Container {
  optionalResult: SearchResult | null | undefined;
  requiredResult: SearchResult;
  optionalResults: SearchResult[] | null | undefined;
  requiredResults: SearchResult[];
}

export type SearchResult = Author | Book;

export type StringOrNumber = string | number;

export type MixedUnion = string | boolean | Author;

export type UnionOfUnions = SearchResult | StringOrNumber;
"
`;
